### DISCUSSION TITLE
message generation does not have have any out files

### DISCUSSION THREAD

**Discussion by LiyouZhou** [mvukov/rules_ros2#34] [Q&A]
I am trying to get the location where I can include headers in a genrule.

So I do "$(location @ros2_rcl_interfaces//:c_builtin_interfaces)", however this results in the following error:

`label '@ros2_rcl_interfaces//:c_builtin_interfaces' in $(location) expression expands to no files`,

@mvukov can you help?

**Comment by mvukov** [comment]
Can you give a bit more context what are u trying to achieve? Indeed, C/C++ code generation rules don't provide DefaultInfo/[OutputGroupInfo](https://bazel.build/rules/lib/OutputGroupInfo) at the moment -- I didn't see a need for that. I could add a depset of headers e.g. in OutputGroupInfo. Then you could process that e.g. in your rule.

BTW, I tend to avoid genrule as it works in the host config. I would try something like https://github.com/bazelbuild/bazel-skylib/blob/main/docs/run_binary_doc.md#run_binary or a custom rule.

**Reply by LiyouZhou** [reply]
I'm trying to get [r2r](https://github.com/sequenceplanner/r2r) to build in bazel. It needs the message type headers to generate bindings in rust. I am using genrule to run the binding generator.  However I am unable to extract the location of the headers from the `c_ros2_interface_library`. This is a quick hack as I need to get something to work quickly. However I am green at bazel, if you know of a simpler way, please let me know.

**Reply by mvukov** [reply]
In principle CGeneratorAspectInfo has CcInfo that holds compilation_context that has all the headers. But, CGeneratorAspectInfo could also be extended to just have the generated headers -- if necessary. Then I would run r2r in another aspect (in aspect-on-aspect fashion) that would use data from CGeneratorAspectInfo and then generate the rust code.
